kafka_consumer/datadog_checks/kafka_consumer/client.py

```diff
@@ -150,99 +153,77 @@ def get_consumer_offsets(self):
+                    if self.config._monitor_unlisted_consumer_groups or not self.config._consumer_groups_compiled_regex:
+                        if offset != OFFSET_INVALID:
+                            consumer_offsets[(consumer_group, topic, partition)] = offset
+                    else:
+                        to_match = f"{consumer_group},{topic},{partition}"
+                        if self.config._consumer_groups_compiled_regex.match(to_match):
+                            if offset != OFFSET_INVALID:
+                                consumer_offsets[(consumer_group, topic, partition)] = offset
 
         return consumer_offsets
 
     def _get_consumer_groups(self):
-        # Get all consumer groups
+        # Get all consumer groups to monitor
         consumer_groups = []
-        consumer_groups_future = self.kafka_client.list_consumer_groups()
-        self.log.debug('MONITOR UNLISTED CG FUTURES: %s', consumer_groups_future)
-        try:
-            list_consumer_groups_result = consumer_groups_future.result()
-            self.log.debug('MONITOR UNLISTED FUTURES RESULT: %s', list_consumer_groups_result)
+        if self.config._monitor_unlisted_consumer_groups or self.config._consumer_groups_compiled_regex:
+            consumer_groups_future = self.kafka_client.list_consumer_groups()
+            try:
+                list_consumer_groups_result = consumer_groups_future.result()
+
+                consumer_groups.extend(
+                    valid_consumer_group.group_id for valid_consumer_group in list_consumer_groups_result.valid
+                )
+            except Exception as e:
+                self.log.error("Failed to collect consumer groups: %s", e)
+            return consumer_groups
+        else:
+            return self.config._consumer_groups
 
-            consumer_groups.extend(
-                valid_consumer_group.group_id for valid_consumer_group in list_consumer_groups_result.valid
-            )
-        except Exception as e:
-            self.log.error("Failed to collect consumer groups: %s", e)
-        return consumer_groups
+    def _list_consumer_group_offsets(self, cg_tp):
+        return self.kafka_client.list_consumer_group_offsets([cg_tp])
 
     def _get_consumer_offset_futures(self, consumer_groups):
-        topic_metadata = self.kafka_client.list_topics(timeout=self.config._request_timeout).topics
-        topics = {
-            topic: list(topic_metadata[topic].partitions.keys())
-            for topic in topic_metadata
-            if topic not in KAFKA_INTERNAL_TOPICS
-        }
-
-        for consumer_group in consumer_groups:
-            self.log.debug('CONSUMER GROUP: %s', consumer_group)
-
-            for topic_partition in self._get_topic_partitions(topics, consumer_group):
-                yield self.kafka_client.list_consumer_group_offsets(
-                    [ConsumerGroupTopicPartitions(consumer_group, [topic_partition])]
-                )[consumer_group]
```

Comment:
I've been looking at this piece of code and I figured that this has something to do with the issue https://github.com/DataDog/integrations-core/issues/15288. Seems like in the DD7.45/7.46 implementation, each consumer group is "associated" with all topics found `self.kafka_client.list_topics(timeout=self.config._request_timeout).topics`, which is wrong and basically makes the check useless. At least in our environment the metric number shot up from 600 to 70k.

Seems like new new implementation 7.47 is going to fix this; can you please confirm this?

Reply:
cc @yzhan289 🙏🏻

Reply:
I found the reasoning and fix https://github.com/DataDog/integrations-core/pull/15098 🎉 .